### PR TITLE
Cpu

### PR DESCRIPTION
**Pull request by billy000400**
Since not every user has a GPU and CUDA version depends on GPU models and OS, I added some cpu support by replacing `.cuda()` to `.to(device)` in models.py and train_model.py, in which the default device is set to be `cpu`. If users want to train the model on GPU, they could set the device variable following the torch converntion, e.g., "cuda:0", install the appropriate CUDA driver & toolkit, and compile torch with CUDA enabled. 

Also, there is an error induced by the `tmul` function in models.py. The traceback is shown below:
```
Traceback (most recent call last):
  File "/home/billyli/UCSD/cookiecutter_bench/hbb_interaction_network/src/models/train_model.py", line 412, in <module>
    main(args)
  File "/home/billyli/UCSD/cookiecutter_bench/hbb_interaction_network/src/models/train_model.py", line 278, in main
    out = gnn(trainingv.to(device), trainingv_sv.to(device))
  File "/home/billyli/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/billyli/UCSD/cookiecutter_bench/hbb_interaction_network/src/models/models.py", line 100, in forward
    Orr = self.tmul(x, self.Rr)
  File "/home/billyli/UCSD/cookiecutter_bench/hbb_interaction_network/src/models/models.py", line 187, in tmul
    return torch.mm(x.view(-1, x_shape[2]), y).view(-1, x_shape[1], y_shape[1])
RuntimeError: mat1 and mat2 shapes cannot be multiplied (7680x30 and 60x3540)
```
Could you make some suggestions on how to solve this error?

### CODE REVIEW

**Comment by jmduarte** [comment]
hi @billy000400 it looks ok to me and i think I've seen this error before but i don't remember the solution...

one thing we want to be careful about is changing the model too much. we want to make sure you can still load the saved state dictionaries that are included in the repo. Could you check if that still works?

File: src/models/predict_model.py [hunk 1]
```diff
@@ -170,20 +177,63 @@ def main(args, save_path="", evaluating_test=True):  # noqa: C901
...
-    for j in tqdm.tqdm(range(0, target_test.shape[0], batch_size)):
-        out_test = softmax(gnn(torch.from_numpy(test[j : j + batch_size]).cuda()))
-        out_test = out_test.cpu().data.numpy()
-        if j == 0:
-            prediction = out_test
-        else:
-            prediction = np.concatenate((prediction, out_test), axis=0)
-        del out_test
+    else:
+        print("22")
```
Comment (jmduarte) [inline src/models/predict_model.py]: why is this here?

File: src/models/predict_model.py [hunk 2]
```diff
@@ -149,13 +154,15 @@ def main(args, save_path="", evaluating_test=True):  # noqa: C901
...
+    test_sv = test_3
     params = params_2
     params_sv = params_3
 
     vv_branch = args.vv_branch
+    set_onnx = args.set_onnx
 
     prediction = np.array([])
-    batch_size = 1024
+    batch_size = 1000  # 1024
```
Comment (jmduarte) [inline src/models/predict_model.py]: maybe this should be a command line argument with a default of 1024 (or 2048 or whatever)?

File: src/models/predict_model.py [hunk 3]
```diff
@@ -24,6 +26,9 @@
 
 
 def main(args, save_path="", evaluating_test=True):  # noqa: C901
+    device = args.device
```
Comment (jmduarte) [inline src/models/predict_model.py:34]: what if we do something like this instead:
```python
device = torch.device("cuda" if torch.cuda.is_available() else "CPU")
```

we should also keep the `import setGPU` but maybe check if `torch.cuda.is_available()` first

File: src/models/predict_model.py [hunk 4]
```diff
@@ -3,10 +3,12 @@
 from pathlib import Path
 
 import numpy as np
-import setGPU  # noqa: F401
```
Comment (jmduarte) [inline src/models/predict_model.py:8]: keep the `import setGPU` but maybe check if `torch.cuda.is_available()` first

File: src/models/predict_model.py [hunk 5]
```diff
@@ -170,20 +177,63 @@ def main(args, save_path="", evaluating_test=True):  # noqa: C901
...
+            dummy_input_2 = test_sv[i : i + batch_size]
+
+            # Load the ONNX model
+            model = onnx.load(model_path)
+
+            # Check that the IR is well formed
+            onnx.checker.check_model(model)
+
+            # Print a human readable representation of the graph
+            # print(onnx.helper.printable_graph(model.graph))
```
Comment (jmduarte) [inline src/models/predict_model.py]: remove commented code

File: src/models/predict_model.py [hunk 6]
```diff
@@ -170,20 +177,63 @@ def main(args, save_path="", evaluating_test=True):  # noqa: C901
...
+            options = ort.SessionOptions()
+            options.intra_op_num_threads = 1
+            ort_session = ort.InferenceSession(model_path, options, providers=[("CUDAExecutionProvider")])
+
+            # compute ONNX Runtime output prediction
+            ort_inputs = {ort_session.get_inputs()[0].name: dummy_input_1, ort_session.get_inputs()[1].name: dummy_input_2}
+            ort_outs = ort_session.run(None, ort_inputs)
+
+            temp_onnx_res = ort_outs[0]
+            #             softmax = torch.nn.Softmax(dim=1)
```
Comment (jmduarte) [inline src/models/predict_model.py]: remove commented code

File: src/models/train_model.py [hunk 7]
```diff
@@ -7,7 +7,6 @@
 from pathlib import Path
 
 import numpy as np
-import setGPU  # noqa: F401
```
Comment (jmduarte) [inline src/models/train_model.py:10]: keep (see comment in `predict_model.py`)

File: src/models/train_model.py [hunk 8]
```diff
@@ -17,9 +16,15 @@
 from src.data.h5data import H5Data
 from src.models.models import GraphNet
 
+# import sys
```
Comment (jmduarte) [inline src/models/train_model.py]: remove commented code

File: src/models/train_model.py [hunk 9]
```diff
@@ -28,16 +33,19 @@
...
 n_targets = len(defn["reduced_labels"])  # number of classes
-device = "cpu"
 params = defn["features_2"]
 params_sv = defn["features_3"]
 
 
 def main(args):  # noqa: C901
     """Main entry point of the app"""
-
+    print(args.random_split)
```
Comment (jmduarte) [inline src/models/train_model.py]: remove

File: src/models/train_model.py [hunk 10]
```diff
@@ -28,16 +33,19 @@
...
     model_dict = {}
 
+    device = args.device
```
Comment (jmduarte) [inline src/models/train_model.py:48]: try approach i commented on in `predict_model.py`

File: src/models/train_model.py [hunk 11]
```diff
@@ -81,29 +90,60 @@ def main(args):  # noqa: C901
...
+            spectators_name="spectator_subgroup",
+        )
+        data_val.set_file_names(files_val)
+
+        n_val = data_val.count_data()
+        n_train = data_train.count_data()
+        print("val data:", n_val)
+        print("train data:", n_train)
+    else:
+        t_X1_tr = np.load("//grand/RAPINS/ruike/npy_data1/data_X1_tr.npy")
```
Comment (jmduarte) [inline src/models/train_model.py]: why is this location hard coded?

**Review by jmduarte — CHANGES_REQUESTED** [review]

File: src/models/predict_model.py (hunk 3 above)
Comment (billy000400) [reply src/models/predict_model.py:34]: what if a user has multiple GPUs and he/she/they wants to specify which one to use? Should we give user this option?

File: src/models/predict_model.py (hunk 1 above)
Comment (billy000400) [reply src/models/predict_model.py]: I rebased my branch on Ruike's branch. Sorry for the confusion :<

**Comment by jmduarte** [comment]
@billy000400 can you update this PR?

@yorkiva is this consistent with the changes you need for inference on CPU vs GPU?

**Comment by yorkiva** [comment]
I have tested the existing models.py script on my laptop. It works fine and
also can import pretrained models.

**Comment by billy000400** [comment]
@jmduarte updated. Lint complained about setGPU is never used. Do we need to change it?
